datafusion/physical-plan/src/work_table.rs

```diff
@@ -169,7 +169,7 @@ impl ExecutionPlan for WorkTableExec {
         &self.cache
     }
 
-    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+    fn children(&self) -> Vec<&Arc<dyn ExecutionPlan>> {
```

Comment:
After reading through this I think it is also an API change (as it changes the API for `ExecutionPlan::children`  -- however I think it is for the best.

Though I wonder if we are going to be changing the signature anyways, I wonder if we should consider something that doesn't require an allocation like

```rust
    fn children(&self) -> [&Arc<dyn ExecutionPlan>] {
```

Instead of 
```rust
    fn children(&self) -> Vec<&Arc<dyn ExecutionPlan>> {
```

🤔

Reply:
I don't think we can return an array if its size is not fixed.

Reply:
Maybe `&[&Arc<dyn ExecutionPlan>]` was meant?

Reply:
I think we could return a slice, but the current `Vec` is in sync with other implementations how we usually return childrens. Like `LogicalPlan::inputs()` or `ConcreteTreeNode::children()`.

Reply:
Actualy, I don't think we can return a slice of references. Returning an empty slice here would be ok, but at other places where there are childrens to return (e.g. in `BinaryExpr`) we need to build container (vec or array) to store the references and then return a slice of the container, but who will own the container?

Reply:
I opened a PR as returning a slice in `ConcreteTreeNode` is possible: https://github.com/apache/datafusion/pull/10666
But it will not work for `LogicalPlan` or `ExecutionPlan` or `PhysicalExpr`.

Reply:
I've tried to return `&[Arc<dyn ExecutionPlan>]` and it fails unsurprisingly -- unless we stores all children in a vector like `Union`:
```rust
    fn children(&self) -> &[Arc<dyn ExecutionPlan>] {
        self.inputs.as_slice()
    }
```
Otherwise we cannot return a temporary slice